ERCS/erc-7820.md

```diff
@@ -0,0 +1,411 @@
+
+### Role Checking
+
+**getRoleInfo(address targetContract, address account, bytes32 role)**: This view function allows the verification of whether a particular account holds a specific role for a given contract. This is essential for ensuring that operations requiring specific permissions are performed only by authorized users.
+
+### Contract Information Retrieval
+
+**getContractInfo(address _contract)**: This function provides the ability to retrieve the status and admin information of a registered contract. It enhances transparency and allows administrators and users to easily query the status and management of any contract within the registry.
+
+### Events
+
+**ContractRegistered(address indexed _contract, address indexed _admin)**: Emitted when a new contract is registered, this event ensures that there is a public record of contract registrations, facilitating auditability and transparency.
+
+**ContractUnregistered(address indexed _contract, address indexed _admin)**: Emitted when a contract is unregistered, this event serves to notify the system and its users of the removal of a contract from the registry, which is critical for maintaining an up-to-date and accurate registry.
+
+**RoleGranted(address indexed targetContract, bytes32 indexed role, address indexed account)**: Emitted when a role is granted to an account, this event provides a public log that can be used to track role assignments and changes, ensuring that role grants are transparent and verifiable.
+
+**RoleRevoked(address indexed targetContract, bytes32 indexed role, address indexed account)**: Emitted when a role is revoked from an account, this event similarly ensures that role revocations are publicly logged and traceable, supporting robust access control management.
+
+### Design Decisions
+There are a few design decisions that have to be explicitly specified to ensure the functionality, security, and efficiency of the IAccessControlRegistry:
+
+#### Decentralized Contract Registration
+
+**No Central Owner**: There is no central owner who can register contracts. This design choice promotes decentralization and ensures that individual contracts are responsible for their own registration and management.
+
+#### Efficient Storage and Lookup
+
+**Mapping Utilization**: The use of mappings for storing contract information (mapping(address => ContractInfo) private contracts) and role assignments (mapping(address => mapping(address => mapping(bytes32 => bool))) private _contractRoles) ensures efficient storage and lookup. This is crucial for maintaining performance in a large-scale system with numerous contracts and roles.
```

Comment:
This is specific to your reference implementation. The rationale section is for documenting the specification itself. If you'd like to explain choices made about your reference implementation, please put it in that section instead.